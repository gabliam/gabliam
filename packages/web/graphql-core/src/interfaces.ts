import { ApolloServerPluginLandingPageGraphQLPlaygroundOptions } from 'apollo-server-core';
import { GraphQLResolveInfo } from 'graphql';

export type extractArgsFn = (
  source: any,
  args: any,
  context: any,
  info: GraphQLResolveInfo
) => any;

export interface GraphqlConfig {
  endpointUrl: string;
  playground: ApolloServerPluginLandingPageGraphQLPlaygroundOptions;
  playgroundEnabled: boolean;
}
