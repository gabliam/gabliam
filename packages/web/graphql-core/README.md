[![Build Status][build-image]][build-url]
[![NPM version][npm-image]][npm-url]

# graphql-core

Package core for gabliam and graphql


# License

  MIT

[build-image]: https://img.shields.io/travis/gabliam/gabliam/master.svg?style=flat-square
[build-url]: https://travis-ci.org/gabliam/gabliam
[npm-image]: https://img.shields.io/npm/v/@gabliam/graphql-core.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@gabliam/graphql-core
