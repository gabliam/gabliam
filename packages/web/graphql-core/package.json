{
  "name": "@gabliam/graphql-core",
  "version": "0.0.0-PLACEHOLDER",
  "description": "Graphql core for gabliam",
  "keywords": [
    "gabliam",
    "plugin",
    "deocrators",
    "graphql"
  ],
  "main": "index.js",
  "typings": "index.d.ts",
  "module": "index.js",
  "scripts": {
    "build": "rimraf lib && tsc -p tsconfig.build.json"
  },
  "files": [
    "lib",
    "src",
    "index.js",
    "index.d.ts"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/gabliam/gabliam.git"
  },
  "author": "David Touzet <dtouzet@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/gabliam/gabliam/issues"
  },
  "homepage": "https://github.com/gabliam/gabliam/tree/master/packages/web/graphql-core#readme",
  "peerDependencies": {
    "@gabliam/core": "0.0.0-PLACEHOLDER"
  },
  "devDependencies": {
    "@gabliam/core": "0.0.0-PLACEHOLDER",
    "@types/debug": "^4.1.5",
    "@types/graphql": "^14.2.3",
    "@types/lodash": "^4.14.136",
    "@types/node": "^12.7.1"
  },
  "dependencies": {
    "apollo-server-core": "^2.8.1",
    "caller": "^1.0.1",
    "debug": "^4.1.1",
    "graphql": "^14.5.0",
    "lodash": "^4.17.15",
    "tslib": "^1.10.0",
    "type-graphql": "^0.17.5"
  },
  "engines": {
    "node": ">=8.9.0"
  },
  "gitHead": "ec328fa0b62f474d470f74a2fc9637838f54091e"
}
