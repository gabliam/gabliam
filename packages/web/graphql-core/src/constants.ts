/** Type in registry */
export const TYPE = {
  Controller: 'GraphqlResolver',
};

export const GRAPHQL_CONFIG = Symbol('GABLIAM/GRAPHQL_PLUGIN_CONFIG');

export const DEBUG_PATH = 'Gabliam:Plugin:GraphqlCore';

export const DEFAULT_END_POINT_URL = '/graphql';

export const APPOLLO_SERVER_BASE = Symbol('APOLLO_SERVER_BASE');
