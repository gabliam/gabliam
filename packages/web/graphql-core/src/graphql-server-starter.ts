import {
  Container,
  Registry,
  Config,
  OnMissingBean,
  Bean,
} from '@gabliam/core';
import {
  APP,
  SERVER,
  ServerStarter,
  WebPluginConfig,
  WEB_PLUGIN_CONFIG,
  SERVER_STARTER,
} from '@gabliam/web-core';
import { ApolloServerBase } from 'apollo-server-core';
import { createServer } from 'http';

@Config(90)
export class GraphqlServerStarterConfig {
  @OnMissingBean(SERVER_STARTER)
  @Bean(SERVER_STARTER)
  createGraphqlServerStarter() {
    return new GraphqlServerStarter();
  }
}

export class GraphqlServerStarter implements ServerStarter {
  apolloServer: ApolloServerBase;

  start(container: Container, registry: Registry) {
    const restConfig = container.get<WebPluginConfig>(WEB_PLUGIN_CONFIG);
    const app = container.get<any>(APP);
    const port = restConfig.port;

    const server = createServer(<any>app);
    server.listen(port, restConfig.hostname);
    server.on('error', onError);
    server.on('listening', onListening);

    this.apolloServer.installSubscriptionHandlers(server);

    container.bind(SERVER).toConstantValue(server);

    /* istanbul ignore next */
    function onError(error: NodeJS.ErrnoException): void {
      // tslint:disable-next-line:curly
      if (error.syscall !== 'listen') throw error;
      const bind = typeof port === 'string' ? 'Pipe ' + port : 'Port ' + port;
      switch (error.code) {
        case 'EACCES':
          console.error(`${bind} requires elevated privileges`);
          process.exit(1);
          break;
        case 'EADDRINUSE':
          console.error(`${bind} is already in use`);
          process.exit(1);
          break;
        default:
          throw error;
      }
    }

    /* istanbul ignore next */
    function onListening(): void {
      const addr = server.address();
      let bind = '';
      if (typeof addr === 'string') {
        bind = `pipe ${addr}`;
      } else if (addr && addr.port) {
        bind = `port ${addr.port}`;
      }
      console.log(`Listening on ${bind}`);
    }
  }
}
