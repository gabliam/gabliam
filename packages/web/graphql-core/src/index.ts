export { TYPE } from './constants';
export * from './graphql-core-plugin';
export * from './interfaces';
export * from './resolver-decorator';
export {
  Arg,
  ArgOptions,
  Args,
  ArgsType,
  Authorized,
  createParamDecorator,
  createMethodDecorator,
  Ctx,
  Directive,
  Extensions,
  registerEnumType,
  Field,
  FieldOptions,
  FieldResolver,
  Info,
  InputType,
  InputTypeOptions,
  InterfaceType,
  InterfaceTypeOptions,
  Mutation,
  ObjectType,
  ObjectTypeOptions,
  PubSub,
  Query,
  Resolver,
  Root,
  Subscription,
  SubscriptionOptions,
  createUnionType,
  UseMiddleware,
  ID, Int, ClassType, Publisher
} from 'type-graphql';
