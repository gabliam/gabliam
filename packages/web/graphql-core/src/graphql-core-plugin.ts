import { Container, GabliamPlugin, Registry, Scan } from '@gabliam/core';
import { WebConfiguration } from '@gabliam/web-core';
import { ApolloServerBase } from 'apollo-server-core';
import { GraphQLSchema } from 'graphql';
import { buildSchema } from 'type-graphql';
import { GRAPHQL_CONFIG, TYPE } from './constants';
import { GraphqlConfig } from './interfaces';

@Scan()
export abstract class GraphqlCorePlugin implements GabliamPlugin {
  async build(container: Container, registry: Registry) {
    const graphqlPluginConfig = container.get<GraphqlConfig>(GRAPHQL_CONFIG);
    const controllerIds = registry.get(TYPE.Controller).map(val => val.target);
    const schema = await buildSchema({
      resolvers: controllerIds as [Function, ...Function[]],
      container: {
        get(someClass) {
          return container.get(someClass);
        },
      },
    });

    const apolloServer = this.getApolloServer(
      container,
      registry,
      graphqlPluginConfig,
      schema
    );

    container.get(WebConfiguration).addServerConfigs(server => {
      apolloServer.installSubscriptionHandlers(server);
      return server;
    });
  }

  abstract getApolloServer(
    container: Container,
    registry: Registry,
    graphqlPluginConfig: GraphqlConfig,
    schema: GraphQLSchema
  ): ApolloServerBase;
}
