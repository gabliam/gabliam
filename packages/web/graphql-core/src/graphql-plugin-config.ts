import { Bean, Joi, PluginConfig, Value } from '@gabliam/core';
import { ApolloServerPluginLandingPageGraphQLPlaygroundOptions } from 'apollo-server-core';
import d from 'debug';
import { DEBUG_PATH, DEFAULT_END_POINT_URL, GRAPHQL_CONFIG } from './constants';
import { GraphqlConfig } from './interfaces';

const debug = d(`${DEBUG_PATH}:GraphqlPluginConfig`);

export const GraphiqlOptionsValidator = Joi.object().keys({
  endpointURL: Joi.string(),
  subscriptionsEndpoint: Joi.string(),
  query: Joi.string(),
  variables: Joi.object(),
  operationName: Joi.number(),
  result: Joi.object(),
  passHeader: Joi.string(),
});

@PluginConfig()
export class GraphqlPluginConfig {
  @Value('application.graphql.endpointUrl', Joi.string())
  private endpointUrl: string = DEFAULT_END_POINT_URL;

  @Value('application.graphql.playground.enabled', Joi.boolean())
  private playgroundEnabled = false;

  @Value('application.graphql.playground.config', GraphiqlOptionsValidator)
  private playground: ApolloServerPluginLandingPageGraphQLPlaygroundOptions;

  @Bean(GRAPHQL_CONFIG)
  creatreConfig(): GraphqlConfig {
    const graphqlConfig: GraphqlConfig = {
      endpointUrl: this.endpointUrl,
      playground: this.playground,
      playgroundEnabled: this.playgroundEnabled,
    };
    debug('GraphqlConfig', graphqlConfig);

    return graphqlConfig;
  }
}
