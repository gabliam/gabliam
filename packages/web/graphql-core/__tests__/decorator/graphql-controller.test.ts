// tslint:disable:one-line
// tslint:disable:no-unused-expression
import { reflection } from '@gabliam/core';
import { GabResolver } from '@gabliam/graphql-core';
import '../../src';
test('should add Register metadata to a class when decorated with @Resolver()', () => {
  @GabResolver()
  class TestBean {}

  expect(reflection.annotations(TestBean)).toMatchSnapshot();
});
