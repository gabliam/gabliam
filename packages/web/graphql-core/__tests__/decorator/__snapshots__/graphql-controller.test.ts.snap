// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add Register metadata to a class when decorated with @Resolver() 1`] = `
Array [
  DecoratorFactory {
    "autobind": true,
    "id": [Function],
    "type": "GraphqlResolver",
  },
]
`;
